### ISSUE TITLE
Clarification about Hyperparameters and CoLar stability

### ISSUE THREAD

**Issue by mohammed-elkomy** [xiaomi-research/colar#12]
Hi @AlbertTan404  

Thanks for the awesome work. I reviewed the code and the different training settings, and I trained a Llama 1b-instruct using CoT to replicate some CoLar results. I used GSM8K datasets; my CoT model has around 40% val acc.

However, I tried to use the math Colar SFT parameters for math you shared here.
https://huggingface.co/AlbertTan/CoLaR/blob/main/logs/colar/qsa-math/colar-math-sft-llama/hparams.yaml

Unfortunately, CoLar is unstable during training and stays at random guess performance even after 10 epochs. 

Could you provide some guidelines on setting the parameters of the latent head and loss relative weights?

These are my current parameters. 
latent_cot_config:
ce_weight: 1
embed_modeling_weight: 1
embed_modeling_loss: mse  # {nll, mse}
entropy_weight: -1e-6     
pred_embed_forward_weight: 1   
max_compression_factor: 5
pred_compressed_cot: True
sqrt_mean: True

GSM8k 
CoT checkpoint as initialization. 

Thank you in advance.

**Comment by AlbertTan404** [comment]
Hi @mohammed-elkomy,

Could you please provide the loss/acc plot and the logged validation generations?
The config is good so I suppose there might be some bugs during training.

There are some of loss plots of mine: https://github.com/xiaomi-research/colar/issues/5#issuecomment-3615874933

**Comment by AlbertTan404** [comment]
@mohammed-elkomy And also some samples from train/val data. Data formatting is crucial for training.

**Comment by mohammed-elkomy** [comment]
@AlbertTan404 Thanks a lot for the quick reply. It seems it's not a configuration issue. I tried a toy dataset of 32 samples and used it for train/test/val (we should expect a perfect performance). However, I got around 3% performance on Colar. Cot is fine-tuning correctly, but other latent models are not.

I will investigate this a bit more, but most probably it's a numerical stability issue or a bug. FYI, I installed a slightly higher version of PyTorch and Python. Is downgrading crucial, as you mentioned in Readme.md? 

APPRECIATED!

**Comment by AlbertTan404** [comment]
Hi @mohammed-elkomy ,
I suggest that you could inspect the training from these perspectives (from the most important to less):
1. Checking whether the input text is in correct format (user prompt + latent generation)
2. Downgrading numpy < 2.0 since we have observed severe numerical instability with np>2.0 (not sure whether np is the root cause). For transformers/pytorch, we do not expect version incompatibility (If you find that it's a pytorch/transoformers version issue, please tell us)
3. Using pred_embed_forward_weight=0 (reduces training complexity)
4. Using max_compression_factor=2 or even 1 (helps convergence)
5. Using entropy_weight=0 (vulnerable to numerical instability)

The Latent Reasoning train&val process could be noisy in the initial stage, since there's a large domain gap between input embeddings and predicted latents. Thus, it requires a full-dataset training for at least 5 epochs (since we randomly sample c=1 to c=5 at each training step) to see the convergence.

If you have any updates or the issue remains, feel free to contact!

**Comment by mohammed-elkomy** [comment]
I tried downgrading numpy and am still experimenting with a toy dataset to test convergence. Generally, if the code is numerically stable, it should get a high score. 

CoT => 99% for the toy gsm8k
Coconut => 40% for the toy gsm8k
CoLaR =>  (3%) still producing random tokens, although the loss got decreased; please check the attached screenshot. **Probably the latent generation logic has some bug.** 

<img width="1170" height="676" alt="Image" src="https://github.com/user-attachments/assets/5f73500e-fc52-4388-83a9-f96c7605ec7d" />
Some additional details:
Here is a sample output of groundtruth and prediction by CoLaR after 90 epochs; it's producing text rather than te answer. 

**Is this a meaningful behavior?**

```
Comparing answers: gt=5, pred=let h be the number of hours it takes joy to read 120 pages                                                                                                                          
Comparing answers: gt=16, pred=roque walks to and from work 3 times a week and each                                                                                                                                
Comparing answers: gt=16, pred=roque walks to work for 2 x 3 = <<2*                                                                                                                                                
Comparing answers: gt=7, pred=bobbie’s name is j-a-m-e-g-r-y so her                                                                                                                                                
Comparing answers: gt=7, pred=s then j for short for samantha then l for full name.                                
```                      
                                                                          
**All predictions are generating max_new_tokens**

Thanks so much for your time. This will be helpful for future collaborators interested in your awesome work.

**Comment by AlbertTan404** [comment]
Hi @mohammed-elkomy. Based on your training log, the root cause is **underfitting**. It would take more steps for converging CoLaR than CoT or Coconut, since we train an extra randomly initialized module (the latent head).

This repo has been used to reproduced CoLaR (see https://github.com/xiaomi-research/colar/issues/5#issuecomment-3960241424). So I would suggest a **full-dataset** run on the gsm dataset till you see the real convergence of ce_loss (<1) and embed_modeling_loss (~0).

If performance issues persist after a full-dataset run, I'm open to an online meeting in Chinese or English :)

**Comment by mohammed-elkomy** [comment]
Hi again @AlbertTan404.

At the beginning, I tried only the original GSM8K due to my limited GPU resources, but as I mentioned, it was numerically unstable. I recently tried GSM8K Aug as you explained in the other thread for 10 epochs, and I now see reasonable performance.
Do you have an idea if this is due to the size of the dataset or maybe the unexpected format? Because GSM8K has text cot samples, while GSM8K-aug uses just equations for the cot. **Do you think this is the reason for my model collapse?** Or let me ask it differently, have you SFT-trained your model with other datasets that have **text cots?** 


Thanks sooooooooooooooo much

**Comment by AlbertTan404** [comment]
Hi @mohammed-elkomy,

We have also trained CoLaR on natural language cots like MATH-500 and it also worked. Thus we think that the main issue is **the number of training iteration**. CoLaR does require more training iterations to fit in a new representation.

A tip for future practice: actually the original cots in the datasets (gsm/gsm-aug/...) are NOT the ideal cots, since they are NOT in the native (to the model itself) reasoning pattern. I would suggest applying a 
```question->model->generated cots by itself->filter by acc->collect filtered cots as training data```
pipeline.

**Comment by hunshi34** [comment]
> Hi [@mohammed-elkomy](https://github.com/mohammed-elkomy),
> 
> We have also trained CoLaR on natural language cots like MATH-500 and it also worked. Thus we think that the main issue is **the number of training iteration**. CoLaR does require more training iterations to fit in a new representation.
> 
> A tip for future practice: actually the original cots in the datasets (gsm/gsm-aug/...) are NOT the ideal cots, since they are NOT in the native (to the model itself) reasoning pattern. I would suggest applying a `question->model->generated cots by itself->filter by acc->collect filtered cots as training data` pipeline.

Hello，do you use the same setting in math500 and gsm-aug? Just like the example in readme file. And is the training of math500 used data generated by model itself? How many epoch do you used in math500?
Looking forward to your reply sincerely.

**Comment by AlbertTan404** [comment]
Hi @hunshi34,

In this repo and our paper, we use the official train/test set and did not apply this data generation pipeline.

If you want to train CoLaR on MATH, we suggest using NLL as the loss function. More details could be found [here](https://huggingface.co/AlbertTan/CoLaR/blob/main/logs/colar/qsa-math/colar-math-sft/hparams.yaml).

**Comment by hunshi34** [comment]
thanks，I will try it

**Comment by mohammed-elkomy** [comment]
Thanks for clarifying @AlbertTan404. I found different versions of the math dataset, and I don't think they offer the same split you used in
[math.py](https://github.com/xiaomi-research/colar/blob/main/data_preprocessing/math.py)
train_ds = train_data[:int(train_length * 0.8)]
val_ds = train_data[int(train_length * 0.8):]
Also, I found some examples in the math dataset having a LaTeX expression; however, verify_answer in [model_base](https://github.com/xiaomi-research/colar/blob/main/src/models/model_base.py) assumes all answers to be just numbers. Could you also clarify this? 

I'm willing to reproduce these results as well, but I believe it will be helpful as well if you could share more details on datasets preprocessing and splitting. (For example the color-math-sft must be tested with exactly the same test dataset you used to eliminate any leakage.)

Thanks again, and all my appreciation for all your support.

**Comment by AlbertTan404** [comment]
Hi @mohammed-elkomy,

We post-processed the MATH dataset from huggingface (see ./data_preprocessing/math.py, there's a relative link to a local json file instead of load_dataset("ankner/math-500")). Using the 80-20 split for train and validation is simply due to they did not provide an official validation set (there are only train and test sets).

Regarding verify_answer: yes, it is a BUG. We origianlly treated the answers in MATH with the same way to GSM. However, some answers in MATH dataset are expressed in latex, so such verification function could introduce many false negatives. In our later practice, we switched to the ```math_verify``` library for more robust answer extraction and verifycation.

Hopefully, I'll re-write the entire repo to make it more robust (probably after my next paper submission ...)

Albert.

**Comment by mohammed-elkomy** [comment]
Thanks for the clarification. I will reproduce these steps too. 
Best of luck with your next submission. You deserve the best. 💪💪
